From a20f2154b2c79eb8550ea9228d1a4415ff51b72a Mon Sep 17 00:00:00 2001
From: Mark Dickinson <mdickinson@enthought.com>
Date: Mon, 4 Mar 2024 13:47:57 +0000
Subject: [PATCH] Fix Traits build error on Python 3.13 (#1767)

This PR updates the core test workflow to run tests on the
in-development 3.13 version of Python, and fixes use of the now-removed
`Py_TRASHCAN_SAFE_BEGIN/END` macros.

Fixes #1765.
---
 traits/ctraits.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/traits/ctraits.c b/traits/ctraits.c
index 6fda770..f4821a2 100644
--- a/traits/ctraits.c
+++ b/traits/ctraits.c
@@ -808,10 +808,18 @@ static void
 has_traits_dealloc(has_traits_object *obj)
 {
     PyObject_GC_UnTrack(obj);
+#if PY_VERSION_HEX < 0x03080000
     Py_TRASHCAN_SAFE_BEGIN(obj);
+#else
+    Py_TRASHCAN_BEGIN(obj, has_traits_dealloc);
+#endif
     has_traits_clear(obj);
     Py_TYPE(obj)->tp_free((PyObject *)obj);
+#if PY_VERSION_HEX < 0x03080000
     Py_TRASHCAN_SAFE_END(obj);
+#else
+    Py_TRASHCAN_END
+#endif
 }
 
 /*-----------------------------------------------------------------------------
@@ -3006,10 +3014,18 @@ static void
 trait_dealloc(trait_object *trait)
 {
     PyObject_GC_UnTrack(trait);
+#if PY_VERSION_HEX < 0x03080000
     Py_TRASHCAN_SAFE_BEGIN(trait);
+#else
+    Py_TRASHCAN_BEGIN(trait, trait_dealloc);
+#endif
     trait_clear(trait);
     Py_TYPE(trait)->tp_free((PyObject *)trait);
+#if PY_VERSION_HEX < 0x03080000
     Py_TRASHCAN_SAFE_END(trait);
+#else
+    Py_TRASHCAN_END
+#endif
 }
 
 /*-----------------------------------------------------------------------------
-- 
2.8.1

